### PR TITLE
Tracking memory usage of `PagePacket` and `RemotePage` in RN fetching pages

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: ref #7670 ref #7628

Problem Summary:  Tracking memory usage of RN fetching pages.

### What is changed and how it works?

- Add `sub_root_of_query_storage_task_mem_trackers`. If a memory tracker of storage tasks is driven by query, it should inherit `sub_root_of_query_storage_task_mem_trackers`.
- Add `fetch_pages_mem_tracker` for tracking memory usage of fetching pages. It inherit from `sub_root_of_query_storage_task_mem_trackers`.
- Tracking memory usage of `PagePacket` and `RemotePage` in fetching pages.
- Add message about memory usage of `sub_root_of_query_storage_task_mem_trackers` and `fetch_pages_mem_tracker` when memory usage exceeded.

image
image

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  - Write a lot of data to TiFlash, make data in `DeltaValueSpace` large.
  - Remove local cache of RNs.
  - Set configuration `profiles.default.max_memory_usage_for_all_queries` to 0.2.
  - Run `select count(*) from t`.
  - TiFlash throws exception of memory limit.
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Common/MemoryTracker.h [hunk 1]
```diff
@@ -157,6 +164,16 @@ extern thread_local MemoryTracker * current_memory_tracker;
 extern std::shared_ptr<MemoryTracker> root_of_non_query_mem_trackers;
 extern std::shared_ptr<MemoryTracker> root_of_query_mem_trackers;
 
+// Initialize in `initStorageMemoryTracker`.
+// If a memory tracker of storage tasks is driven by query, it should inherit `sub_root_of_query_storage_task_mem_trackers`.
+// root_of_query_mem_trackers
+//             |-- sub_root_of_query_storage_task_mem_trackers
+//                                    |-- rn_fetch_pages_mem_tracker
+extern std::shared_ptr<MemoryTracker> sub_root_of_query_storage_task_mem_trackers;
```
Comment (bestwoody) [inline dbms/src/Common/MemoryTracker.h:172]: can we print usage of `sub_root_of_query_storage_task_mem_trackers` and `rn_fetch_pages_mem_tracker` along with the existed err log, when out of quota occur in `memtracker::alloc(size)`

**Review by JaySon-Huang — APPROVED** [review]

File: dbms/src/Common/MemoryTracker.cpp [hunk 2]
```diff
@@ -228,6 +256,22 @@ thread_local MemoryTracker * current_memory_tracker = nullptr;
...
 std::shared_ptr<MemoryTracker> root_of_query_mem_trackers = MemoryTracker::createGlobalRoot();
 
+std::shared_ptr<MemoryTracker> sub_root_of_query_storage_task_mem_trackers;
+std::shared_ptr<MemoryTracker> rn_fetch_pages_mem_tracker;
+
+void initMemoryTracker()
+{
+    RUNTIME_CHECK(sub_root_of_query_storage_task_mem_trackers == nullptr);
+    sub_root_of_query_storage_task_mem_trackers = MemoryTracker::create();
+    sub_root_of_query_storage_task_mem_trackers->setNext(root_of_query_mem_trackers.get());
```
Comment (windtalker) [inline dbms/src/Common/MemoryTracker.cpp:266]: `root_of_query_mem_trackers` will be reset when there is no active queries, I'm not sure if it still ok to be reset after this pr?

Comment (bestwoody) [reply dbms/src/Common/MemoryTracker.cpp:266]: when memtrackers of query_storage and page are reset，are their memory really released？

File: dbms/src/Common/MemoryTracker.cpp [hunk 3]
```diff
@@ -155,6 +171,11 @@ void MemoryTracker::alloc(Int64 size, bool check_memory_limit)
                                   formatReadableSizeWithBinarySuffix(current_limit));
             }
 
+            if (sub_root_of_query_storage_task_mem_trackers)
```
Comment (bestwoody) [inline dbms/src/Common/MemoryTracker.cpp:174]: how about printing `root_of_non_query` , `sub_root_of_query_storage_task_mem_trackers`, `fetch_pages_mem_tracker` in every case.

File: dbms/src/Common/MemoryTracker.cpp [hunk 4]
```diff
@@ -65,6 +71,13 @@ static Poco::Logger * getLogger()
     return logger;
 }
 
+static String memoryUsageDetail()
```
Comment (bestwoody) [inline dbms/src/Common/MemoryTracker.cpp:74]: name is too generic

File: dbms/src/Common/MemoryTracker.cpp [hunk 5]
```diff
@@ -228,6 +259,23 @@ thread_local MemoryTracker * current_memory_tracker = nullptr;
 std::shared_ptr<MemoryTracker> root_of_non_query_mem_trackers = MemoryTracker::createGlobalRoot();
 std::shared_ptr<MemoryTracker> root_of_query_mem_trackers = MemoryTracker::createGlobalRoot();
 
+std::shared_ptr<MemoryTracker> sub_root_of_query_storage_task_mem_trackers;
+std::shared_ptr<MemoryTracker> fetch_pages_mem_tracker;
+
+void initMemoryTracker(Int64 limit, Int64 larger_than_limit)
```
Comment (bestwoody) [inline dbms/src/Common/MemoryTracker.cpp:265]: name is too generic

**Review by bestwoody — APPROVED** [review]
lgtm

**Pull request merged**